src/BucketTesting/Domain/Model/BucketLog.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\BucketTesting\Domain\Model;
+
+use DateTime;
+use DateTimeInterface;
+
+class BucketLog {
+
+	private int $id;
+	private int $external_id;
+	private string $eventName;
+	private array $buckets;
+	private DateTimeInterface $date;
+
+	public function __construct( int $external_id, string $eventName, array $buckets, ?DateTime $date = null ) {
+		$this->external_id = $external_id;
+		$this->eventName = $eventName;
+		$this->date = $date ?? new DateTime();
+
+		$this->buckets = [];
+		foreach ($buckets as $bucket) {
```

Comment:
Please move creation of the `BucketLogBucket` instances to `DatabaseBucketLogger::getBucketMap` and provide the finished array to the constructor. This avoids array-key-to-object mismatches and we stay type-safe the whole way.

Reply:
Hm, there's a circular dependency then as each entity will have the other in their constructors